src/erp12/ga_clj/toolbox.cljc

```diff
@@ -25,12 +25,47 @@
   (fn [population]
     (apply min-key by (take size (shuffle population)))))
 
-;; @todo How to handle downsampled lexicase selection? Where do cases come from?
+
+(defn lexicase-selection
+  "Helper for make-lexicase-selection. Takes the population and cases shuffled
+   in a random order. Behaves deterministically to aid in testing."
+  [candidates cases]
+  ;; Stop when cases is empty or candidates has only 1 left.
+  (if (or (empty? cases)
+          (= 1 (count candidates)))
+    (rand-nth candidates)
+    (let [the-case (first cases)
+          best (apply min (map #(nth (:errors %) the-case)
+                               candidates))]
+      (recur (filter #(= best (nth (:errors %) the-case))
+                     candidates)
+             (rest cases)))))
+
+;; @todo Implement epsilon lexicase
+;; Not sure how where I would calculate epsilons just once per generation.
+(defn epsilon-lexicase-selection
+  "Implements semi-dynamic epsilon lexicase, which seems best in trial.
+   semi-dynamic = use local best, but global epsilons that are calculated
+   only once per generation."
+  [candidates cases]
+  :TODO
+  )
```

Comment:
Non-epsilon lexicase is equivalent to epsilon-lexicase with an epsilon of zero, right? What do you think about using a single function where epsilon is set using some parameter than can either be:
1.  nil or false denoting an epsilon of zero.
2. A number denoting a constant epsilon.
3. The keyword `:semi-dynamic`, `:auto`, or something similar to denote the usual calculation for epsilon.

Reply:
I had a reason to implement them separately, but I didn't write it in a comment, and now I forget why. Let's see if I can remember as I go through the rest of this...

Reply:
I can't remember, so maybe I'll merge them when I do epsilon lexicase.

@erp12 Any idea where I'd put the code to calculate the epsilons just once per generation?

Reply:
Good question. I had to think this over for a while and now have a few thoughts.

1. I'm not sure epsilon can always be computed once per generation. My understanding of semi-dynamic epsilon lexicase is that it computes the per-case epsilons based on the collection of individuals being selected from. We don't know that all instances of lexicase selection will A) be within a "generation" or B) be passed the entire population. It might be less quantum-entanglement if we just recompute epsilon each call and accept the added cost.

2. Regardless, you raise a good point about the need for per-generation data, even if we decide epsilon isn't in it. Perhaps the `finalization` function should return a "context" map with the population in it. This "context" map could be what is passed to the `breed` function. Just a thought. I'm open to ideas.

3. Given that we want to solve (2) anyhow, maybe there could be option for specifying epsilon in lexicase selection. If the user passes in a vector, assume it is pre-computed per-case epsilon values. Users can use our mechanism for per-generation data to avoid recomputing epsilon, if appropriate, and otherwise fall back on computing epsilon each call to lexicase. Does that make sense?

Feel free to pick and choose what you want to implement now versus leave as a todo.

Reply:
So, for 1, you're saying that someone might want to use this epsilon lexicase as part of a non-generational system, or one that runs lexicase on some group of individuals that isn't the entire population? I could see that. That said, I worry that trying to make code like this perfectly general for all possible futures makes it too hard to just get it done in a reasonable time. I would prefer to get some version done, and change or add to it later if those use-cases arise. 

With all of this in mind, my preference would be to get _one_ version of epsilon lexicase done (or maybe even leave it unimplemented), and then work on other stuff that we'll actually use in experiments we need to get running. Putting lots of effort into epsilon lexicase that we aren't going to immediately use seems like putting the cart before the horse.

Reply:
Sure. I think the 2 ways to accomplish this are:

1. Compute epsilon inside the call to lexicase. In other words, not once per generation since no place for that exists yet.
2. Leave epsilon lexicase selection as a todo.

I'm fine with either.

Reply:
Let's leave epsilon lexicase todo.